Use Pytest

nicoddemus opened the pull request:
As discussed in #357.

I suggest to review each commit individually to see the progress. 

Initially I just changed the runner to `pytest` and all tests pass:

```
$ coverage run --source . -m pytest conda_smithy/tests
============================= test session starts ==============================
platform linux -- Python 3.5.2, pytest-3.0.4, py-1.4.31, pluggy-0.4.0
rootdir: /home/travis/build/nicoddemus/conda-smithy, inifile: 
collected 28 items 
conda_smithy/tests/test_configure_feedstock.py ..
conda_smithy/tests/test_feedstock_io.py ......
conda_smithy/tests/test_lint_recipe.py ....................
========================== 28 passed in 6.59 seconds ===========================
```

Next I updated all `self.assert*` methods to plain asserts using [unittest2pytest](pypi.org/project/unittest2pytest).

~~Now I will work on using some of pytest features to simplify the tests.~~ 😁 

Converted all tests to use pytest's features like parametrization and fixtures. 😁

jakirkham commented on setup.py:
```diff
@@ -21,8 +21,6 @@ def main():
         zip_safe=False,
         cmdclass=versioneer.get_cmdclass(),
-        tests_require=['six'],
-        test_suite='conda_smithy',
```
Would it be possible to do something like [this]( http://pytest.org/2.2.4/goodpractises.html#integration-with-setuptools-distribute-test-commands )?

nicoddemus replied:
Yes, we can even use [pytest-runner](https://pypi.python.org/pypi/pytest-runner) if you prefer.

jakirkham replied:
Whichever you think is better. 👍

nicoddemus replied:
I personally prefer to just call `pytest` instead of using `setuptools`'s `setup.py test`... In my experience it does not add much to the experience, plus it seems even `setuptools` itself is [moving away](pypa/setuptools#814) from using `setup.py test`.

@jaraco may have something to add on this matter.

nicoddemus commented:
I'm having a lot of errors when trying to run under windows. One that is particularly to me is `Test_fudge_subdir.test_fetch_index` which is failing with this error:

```
url = 'https://repo.continuum.io/pkgs/msys2/osx-64/', cache_dir = None
use_cache = False
session = <conda.connection.CondaSession object at 0x0724C7F0>

    @dotlog_on_return("fetching repodata:")
    def fetch_repodata(url, cache_dir=None, use_cache=False, session=None):
        if not offline_keep(url):
            return {'packages': {}}
        cache_path = join(cache_dir or create_cache_dir(), cache_fn_url(url))
        try:
            with open(cache_path) as f:
                cache = json.load(f)
        except (IOError, ValueError):
            cache = {'packages': {}}
    
        if use_cache:
            return cache
    
        if not ssl_verify:
            try:
                from requests.packages.urllib3.connectionpool import InsecureRequestWarning
            except ImportError:
                pass
            else:
                warnings.simplefilter('ignore', InsecureRequestWarning)
    
        session = session or CondaSession()
    
        headers = {}
        if "_etag" in cache:
            headers["If-None-Match"] = cache["_etag"]
        if "_mod" in cache:
            headers["If-Modified-Since"] = cache["_mod"]
    
        if 'repo.continuum.io' in url or url.startswith("file://"):
            filename = 'repodata.json.bz2'
        else:
            headers['Accept-Encoding'] = 'gzip, deflate, compress, identity'
            headers['Content-Type'] = 'application/json'
            filename = 'repodata.json'
    
        try:
            resp = session.get(url + filename, headers=headers, proxies=session.proxies)
            resp.raise_for_status()
            if resp.status_code != 304:
                if filename.endswith('.bz2'):
                    json_str = bz2.decompress(resp.content).decode('utf-8')
                else:
                    json_str = resp.content.decode('utf-8')
                cache = json.loads(json_str)
                add_http_value_to_dict(resp, 'Etag', cache, '_etag')
                add_http_value_to_dict(resp, 'Last-Modified', cache, '_mod')
    
        except ValueError as e:
            raise RuntimeError("Invalid index file: {0}{1}: {2}"
                               .format(remove_binstar_tokens(url), filename, e))
    
        except requests.exceptions.HTTPError as e:
            if e.response.status_code == 407:  # Proxy Authentication Required
                handle_proxy_407(url, session)
                # Try again
                return fetch_repodata(url, cache_dir=cache_dir, use_cache=use_cache, session=session)
    
            if e.response.status_code == 404:
                if url.startswith(DEFAULT_CHANNEL_ALIAS):
                    user = remove_binstar_tokens(url) \
                                 .split(DEFAULT_CHANNEL_ALIAS)[1] \
                                 .split("/")[0]
                    msg = 'Could not find anaconda.org user %s' % user
                else:
                    if url.endswith('/noarch/'):  # noarch directory might not exist
                        return None
                    msg = 'Could not find URL: %s' % remove_binstar_tokens(url)
            elif e.response.status_code == 403 and url.endswith('/noarch/'):
                return None
    
            elif (e.response.status_code == 401 and
                    rc.get('channel_alias', DEFAULT_CHANNEL_ALIAS) in url):
                # Note, this will not trigger if the binstar configured url does
                # not match the conda configured one.
                msg = ("Warning: you may need to login to anaconda.org again with "
                       "'anaconda login' to access private packages(%s, %s)" %
                       (hide_binstar_tokens(url), e))
                stderrlog.info(msg)
                return fetch_repodata(remove_binstar_tokens(url),
                                      cache_dir=cache_dir,
                                      use_cache=use_cache, session=session)
    
            else:
                msg = "HTTPError: %s: %s\n" % (e, remove_binstar_tokens(url))
    
            log.debug(msg)
>           raise RuntimeError(msg)
E           RuntimeError: Could not find URL: https://repo.continuum.io/pkgs/msys2/osx-64/
```

That URL indeed leads to a `404` error, I'm puzzled on how it is passing on Travis. @jakirkham any advice?

nicoddemus commented:
Converted all tests to use pytest's features like fixtures and parametrization. Good news is that almost all tests pass on Windows except for `Test_fudge_subdir.test_fetch_index` mentioned above:

```
============================= test session starts =============================
platform win32 -- Python 3.4.4, pytest-3.0.4, py-1.4.31, pluggy-0.4.0
rootdir: X:\conda-smithy, inifile:
plugins: instafail-0.3.0, qt-1.2.3, xdist-1.11
collected 74 items

test_configure_feedstock.py .F
test_feedstock_io.py ....s.s.s.s.............................
test_lint_recipe.py ................................

================================== FAILURES ===================================
```

Are you guys interested in testing this in Windows too? If yes, I can add CI for AppVeyor as well.

nicoddemus commented:
Btw, please feel free to ask me any questions related to the conversion.

nicoddemus commented:
Coverage has reduced because now some modules are not imported by the tests. I guess previously the command `setup.py test` would import all modules implicitly and execute all `unittest.TestCase` subclasses it finds. pytest only collects `test_*.py` files for tests.

**master**

```
$ coverage report
Name                                 Stmts   Miss Branch BrMiss  Cover
----------------------------------------------------------------------
bootstrap-obvious-ci-and-miniconda      52     52     20     20     0%
conda_smithy/__init__                    3      0      0      0   100%
conda_smithy/ci_register               196    164     66     66    12%
conda_smithy/cli                       171    128     36     36    21%
conda_smithy/configure_feedstock       292    239    120    117    14%
conda_smithy/feedstock_io               53      1     14      0    99%
conda_smithy/feedstocks                170    146     53     53    11%
conda_smithy/github                     39     31     13     13    15%
conda_smithy/lint_recipe               102      4     50      1    97%
setup                                    5      0      0      0   100%
----------------------------------------------------------------------
TOTAL                                 1083    765    372    306    26%
```

**pytest**

```
0.29s$ coverage report
Name                               Stmts   Miss Branch BrMiss  Cover
--------------------------------------------------------------------
conda_smithy/__init__                  3      0      0      0   100%
conda_smithy/ci_register             196    196     66     66     0%
conda_smithy/cli                     171    171     36     36     0%
conda_smithy/configure_feedstock     292    239    120    117    14%
conda_smithy/feedstock_io             53      1     14      0    99%
conda_smithy/feedstocks              170    170     53     53     0%
conda_smithy/github                   39     39     13     13     0%
conda_smithy/lint_recipe             102      4     50      1    97%
--------------------------------------------------------------------
TOTAL                               1026    820    352    286    20%
```

If you would like to keep the coverage like before, I guess we can write a single test which imports all modules, but it feels like cheating; we should probably focus on writing tests for the missing modules instead.

nicoddemus commented:
> If you would like to keep the coverage like before, I guess we can write a single test which imports all modules, but it feels like cheating; we should probably focus on writing tests for the missing modules instead.

A third option is to make pytest import all python modules like unittest does, which would bring coverage to the exact same level as before. We just need to add this to `setup.cfg`:

```ini
[tool:pytest]
python_files = *.py
```

nicoddemus commented:
Hi,

Just came back to this and realized that it is not really feasible to merge/rebase because there are too much conflicts. 😬 

In retrospect, I should have kept this PR as simple as possible so it could be reviewed and merged soon. As a maintainer myself, I understand how PRs with too much changes might be daunting to merge and end up in the back-burner.

Regardless, is there still interest of using pytest in conda-smithy? If so, I would be happy to open a new PR from `master` with as few changes as possible, perhaps only changing `self.assert*` methods to plain `assert` statements. If desired, more advanced features might be introduced later in any pace the maintainers are comfortable with.

jakirkham commented:
Sorry @nicoddemus, the tests got heavily revised as we tried to address different subtle issues with `git`, file system permissions, line endings, channel sources, etc. all driven by user demand, which has created all of these conflicts (and likely drawn attention from making the test suite healthier 😢). That said, I think the tests and hopefully `conda-smithy` are in a more stable place (i.e. not likely to change too much). Thus it should be in a good place to do some maintenance and to test more thoroughly to keep it there. 😄

Though I agree with your statement that we should aim to do this in smaller pieces. I'll try to do better on my part by directing you to the parts that aren't too likely to change soon and/or get the most benefit from something like pytest so as to avoid conflicts and/or show a clear improvement. Sorry I did a poor job at that before.

The two places that come to mind are `test_configure_feedstock.py` and `test_feedstock_io.py`. As far as `test_configure_feedstock.py` goes, there is relatively little going on there. So it should be a good testing ground and a quick PR. That being said, I don't know if that demonstrates the full potential of pytest or not. Concerning `test_feedstock_io.py`, that test suite is in a pretty ugly place. It is definitely due for a cleanup and from what I've seen of pytest it would add significant value to have it rewritten with pytest. That being said, it may be a bit of a project depending on what time you have to contribute to the effort. To keep things simple, each of these can be a separate PR to ease merging. 😄

While other parts like the linter tests could benefit from a rewrite, there is discussion about changing the tests there ( https://github.com/conda-forge/conda-smithy/issues/423 ) and has been discussion for some time about refactoring the linter out from `conda-smithy` ( https://github.com/conda-forge/conda-smithy/issues/386 ) ( https://github.com/conda-forge/conda-smithy/issues/61#issuecomment-185398366 ). Increasingly updating the linter and smithy's core features (creating feedstocks) feel at odds when releases come up. So it might be better for us to sort out these particular discussions before potentially getting bogged down. There is some concern of this coming up with any re-rendering tests in the future due to issues like these ( https://github.com/conda-forge/conda-smithy/issues/411 ) ( https://github.com/conda-forge/conda-smithy/issues/415 ) ( https://github.com/conda-forge/conda-smithy/issues/447 ). Of them fixing R is pretty pressing, but should hopefully be pretty minimal in impact (maybe just a 1 line change).

So if you have time and interest to look at `test_configure_feedstock.py` and/or `test_feedstock_io.py`, I think those would be good places to start.

FYI - I don't expect I'll have time to look this week and will be out next week, but should be able to look after I get back. So if I'm a bit unresponsive that is why.

jakirkham commented:
Another thing that I would like to pick your brain on is getting better tests for things like CI registration. I'd like these to work in an offline mode (as they typically use a Web API). Maybe using something like betamax or similar would be the right way to go? We could open a new issue to discuss this instead of this thread though.

nicoddemus commented:
> Sorry I did a poor job at that before.

Please not at all. As I said, I got too greedy with the scope of the PR, adding too many modifications which made it harder than necessary for review/inclusion. No need to apologize! 😄 

Noted about your suggestions, I also thought `test_configure_feedstock.py` was a good candidate.

I will limit my next PR then to:

1. Change runner to `pytest`
2. Update `test_configure_feedstock.py` to use plain asserts and the build-in `tmpdir` fixture which eliminates the need for the custom [`tmp_directory`](https://github.com/conda-forge/conda-smithy/blob/master/conda_smithy/tests/test_configure_feedstock.py#L14) function.

I will leave the other test files untouched, which is serves two purposes 1) make the PR small as possible and 2) prove that `pytest` can run usual normal `unittest`-based tests normally and without modification.

> FYI - I don't expect I'll have time to look this week and will be out next week, but should be able to look after I get back. So if I'm a bit unresponsive that is why.

No problem, thanks for the heads up!
